src/math/curve-iterator.js

```diff
@@ -0,0 +1,225 @@
+         * @description Reset the iterator to the specified time
+         * @param {Number} time The new time for the iterator
+         */
+        reset: function (time) {
+            var keys = this.curve.keys;
+            var len = keys.length;
+
+            this.time_ = time;
+            if (!len) {
+                // curve is empty
+                this.left = -Infinity;
+                this.right = Infinity;
+                this.recip = 0;
+                this.p0 = this.p1 = this.m0 = this.m1 = 0;
+            } else {
+                if (this.time_ < keys[0][0]) {
+                    // iterator falls to the left of the start of the curve
+                    this.left = -Infinity;
+                    this.right = keys[0][0];
+                    this.recip = 0;
+                    this.p0 = this.p1 = keys[0][1];
+                    this.m0 = this.m1 = 0;
+                } else if (this.time_ >= keys[len - 1][0]) {
+                    // iterator falls to the right of the end of the curve
+                    this.left = keys[len - 1][0];
+                    this.right = Infinity;
+                    this.recip = 0;
+                    this.p0 = this.p1 = keys[len - 1][1];
+                    this.m0 = this.m1 = 0;
+                } else {
+                    // iterator falls within the bounds of the curve
+                    // perform a linear search for the key just the to left of the
+                    // current time.
+                    // (TODO: for cases where the curve has more than 'n' keys it will
+                    // be more efficient to perform a binary search here instead. Which is
+                    // straight forward thanks to the sorted list of knots).
+                    var index = 0;
+                    while (this.time_ >= keys[index + 1][0]) {
+                        index++;
+                    }
+                    this.left = keys[index][0];
+                    this.right = keys[index + 1][0];
+                    var diff = 1.0 / (this.right - this.left);
+                    this.recip = (isFinite(diff) ? diff : 0);
+                    this.p0 = keys[index][1];
+                    this.p1 = keys[index + 1][1];
+                    if (this._isHermite()) {
+                        this._calcTangents(keys, index);
+                    }
+                }
+            }
+        },
+
+        // returns true if the curve is a hermite and false otherwise
+        _isHermite: function () {
+            return [pc.CURVE_CATMULL,
+                pc.CURVE_CARDINAL,
+                pc.CURVE_CARDINAL_STABLE].indexOf(this.curve.type) != -1;
+        },
+
+        // calculate tangents for the hermite curve
+        _calcTangents: function (keys, index) {
+            var a;
+            var b = keys[index];
+            var c = keys[index + 1];
+            var d;
+
+            if (index === 0) {
+                a = [keys[0][0] + (keys[0][0] - keys[1][0]),
```

Comment:
Is it possible to avoid making new arrays here for these internal function variables?

Reply:
The two cases where arrays are created should be exception and it means that during the (hopefully) more common cases we reference existing arrays. (Meaning we don't have to make copies of variables in the common case). Is that a reasonable tradeoff?